File: Generated_test_cases/JS/src/jobmanager.js
class JobManager {
    constructor() {
        this.data = [];
        this.cleanups = [];
    }

    addData(item) {
        this.data.push(item);
    }

    dataCount() {
        return this.data.length;
    }

    // VULN-2: Capture of index after slice may reallocate – use of stale reference
    scheduleCleanup(index) {
        if (index >= this.data.length) return;
        const ref = this.data[index];        // reference to string (strings are immutable, but we use as example)
        this.cleanups.push(() => {
            // If this.data has been cleared or reordered, ref is still valid string,
            // but the semantic vulnerability would be if ref was an object that was mutated elsewhere.
            console.log(`Cleaning chunk: ${ref}`);
        });
    }

    runCleanup() {
        for (const fn of this.cleanups) {
            fn();
        }
    }
}

module.exports = JobManager;

File: Generated_test_cases/GO/jobs/jobmanager.go
package jobs

import "fmt"

type cleanupFunc func()

type JobManager struct {
	data          []string
	cleanupFns    []cleanupFunc
}

func NewJobManager() *JobManager {
	return &JobManager{}
}

func (jm *JobManager) AddData(d string) {
	jm.data = append(jm.data, d)
}

func (jm *JobManager) DataCount() int64 {
	return int64(len(jm.data))
}

// VULN-2: Capture pointer to string element that may become invalid after append
func (jm *JobManager) ScheduleCleanup(index int64) {
	if index >= int64(len(jm.data)) {
		return
	}
	ptr := &jm.data[index] // capture address of slice element
	jm.cleanupFns = append(jm.cleanupFns, func() {
		// later, if slice reallocated, ptr is dangling
		fmt.Printf("Cleaning chunk: %s\n", *ptr)
	})
}

func (jm *JobManager) RunCleanup() {
	for _, fn := range jm.cleanupFns {
		fn()
	}
}